From 074e08285223f5cb0c96aff27edbb70df6202369 Mon Sep 17 00:00:00 2001
From: Akansha Verma <akanverm@qti.qualcomm.com>
Date: Fri, 7 Aug 2026 14:33:12 +0530
Subject: [PATCH] [linux-6.6]: Run ECM hook before PON iftrap RX hook

Move the PON iftrap receive handling after the ECM fast receive handling
in the receive path.
This allows packets handled by the ECM fast path to be consumed before
they reach the PON iftrap path, preserving the intended RX hook ordering.

Change-Id: I58ccbf8871feb82c52100869768e7dd242e757d3
Signed-off-by: Akansha Verma <akanverm@qti.qualcomm.com>
---
 net/core/dev.c | 38 ++++++++++++++++++++++++--------------
 1 file changed, 24 insertions(+), 14 deletions(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index 9741433edb99..2b06ee7c88fb 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -4869,8 +4869,13 @@ bool dev_fast_xmit(struct sk_buff *skb,
 
 	rcu_read_lock_bh();
 
+	/*
+	 * A packet that has already been decided to be fast forwarded/transmitted
+	 * has had its Tx/Rx PON editing taken care of on that path already, so
+	 * pon iftrap processing is not required for it.
+	 */
 #ifdef CONFIG_IPQ_PON
-	if (pon_iftrap_process_tx(skb, DEV_FAST_XMIT)) {
+	if (!(skb->fast_forwarded || skb->fast_xmit) && (pon_iftrap_process_tx(skb, DEV_FAST_XMIT))) {
 		rcu_read_unlock_bh();
 		return true;
 	}
@@ -4967,8 +4972,13 @@ int __dev_queue_xmit(struct sk_buff *skb, struct net_device *sb_dev)
 
 	skb_update_prio(skb);
 
+	/*
+	 * A packet that has already been decided to be fast forwarded/transmitted
+	 * has had its Tx/Rx PON editing taken care of on that path already, so
+	 * pon iftrap processing is not required for it.
+	 */
 #ifdef CONFIG_IPQ_PON
-	if (pon_iftrap_process_tx(skb, __DEV_QUEUE_XMIT)) {
+	if (!(skb->fast_forwarded || skb->fast_xmit) && pon_iftrap_process_tx(skb, __DEV_QUEUE_XMIT)) {
 		rc = NETDEV_TX_OK;
 		goto out;
 	}
@@ -6088,6 +6098,18 @@ static int __netif_receive_skb_core(struct sk_buff **pskb, bool pfmemalloc,
 		}
 	}
 
+	if (likely(!fast_tc_filter)) {
+		fast_recv = rcu_dereference(athrs_fast_nat_recv);
+		if (fast_recv) {
+			if (fast_recv(skb)) {
+				ret = NET_RX_SUCCESS;
+				goto out;
+			}
+		}
+	}
+
+	skb_recycler_clear_fast_flags(skb);
+
 #ifdef CONFIG_IPQ_PON
 	if (pon_iftrap_enable) {
 		int (*pon_uni_recv)(struct sk_buff *skb);
@@ -6102,18 +6124,6 @@ static int __netif_receive_skb_core(struct sk_buff **pskb, bool pfmemalloc,
 	}
 #endif
 
-	if (likely(!fast_tc_filter)) {
-		fast_recv = rcu_dereference(athrs_fast_nat_recv);
-		if (fast_recv) {
-			if (fast_recv(skb)) {
-				ret = NET_RX_SUCCESS;
-				goto out;
-			}
-		}
-	}
-
-	skb_recycler_clear_fast_flags(skb);
-
 	if (eth_type_vlan(skb->protocol)) {
 		skb = skb_vlan_untag(skb);
 		if (unlikely(!skb))
-- 
2.34.1

